core/bifrost.go

```diff
@@ -50,6 +50,70 @@ type Bifrost struct {
 	backgroundCtx       context.Context                               // Shared background context for nil context handling
 }
 
+// PluginPipeline encapsulates the execution of plugin PreHooks and PostHooks, tracks which plugins ran, and manages short-circuiting and error aggregation.
+type PluginPipeline struct {
+	plugins []schemas.Plugin
+	logger  schemas.Logger
+
+	// Indices of plugins whose PreHook ran (for reverse PostHook)
+	preHookRan []int
+	// Errors from PreHooks and PostHooks
+	preHookErrors  []error
```

Comment:
usually we can wrap it under a separate struct for better abstraction. Also we can create pool of these objects to avoid any runtime allocations

```go
type Errs struct {
    Errors []errors
}

func (e *Errs) Print() {
}

func (e *Errs) Add(e error) {
}
```

Reply:
yes I've made changes to this, I pushed it to the next pr by mistake

Reply:
I'll add sync pools, even I was thinking about it